keps/sig-node/20190806-memory-manager.md

```diff
@@ -0,0 +1,264 @@
+      - [Provide topology hint for Topology Manager](#provide-topology-hint-for-topology-manager)
+      - [Isolate memory and hugepages for a container](#isolate-memory-and-hugepages-for-a-container)
+      - [New Interfaces](#new-interfaces)
+    - [Topology Manager changes](#topology-manager-changes)
+    - [Internal Container Lifecycle changes](#internal-container-lifecycle-changes)
+    - [Feature Gate and Kubelet Flags](#feature-gate-and-kubelet-flags)
+- [Graduation Criteria](#graduation-criteria)
+  - [Phase 1: Alpha (target v1.1x)](#phase-1-Alpha-target-v11x)
+  - [Phase 2: Beta](#phase-1-Beta)
+  - [GA (stable)](#ga-stable)
+
+# Overview
+
+NUMA Awareness is well known for a solution to boost performance for diverse use cases including DPDK and Database. Kubernetes support NUMA related features for NUMA sensitive containers.
+The features(CPU Manager, Device Manager, Topology Manager) treat CPU, Device and it's NUMA topology, but there's no feature for memory and hugepages. It can cause inter-NUMA node communication for Memory access which leads to increase an I/O latency and decrease performance.
+
+The Memory Manager is proposed for a solution deploying Pod and Containers with Guaranteed QoS class of Memory and Hugepages under NUMA awareness.
+
+# Motivation
+
+## Related Features
+
+- [Node Topology Manager][node-topology-manager] is a feature that collects topology hint from various hint providers to calculate socket affinity for a container. Topology Manager judge container whether admit or not under configured policy and socket affinity.
+
+- [CPU Manager][cpu-manager] is a feature provides a solution for CPU Pinning based on cgroups cpuset subsystem, it also offer topology hint to Topology Manager.
+
+- [Device Manager][device-manager] is a one of features provides topology hint for Topology Manager. The main objectivity of Device Manager is allow vendors to adverise their resources to kubelet.
+
+- [Hugepages][hugepages] is a feature allows container to consume pre-allocated hugepages with pod isolation of huge pages.
+
+- [Node Allocatable Feature][node-allocatable-feature] is a feature helps to reserve compute resources to prevent resource starvation. The kube-reserved and system-reserved is used to reserve resources for kubelet and system(OS, etc). Now following resources are supported to reserve: cpu, memory, ephemral storage.
+
+## Related issues
+
+- [Hardware topology awareness at node level (including NUMA)][numa-issue]
+- [Support Container Isolation of Hugepages][hugepage-issue]
+- [Support isolating memory to single NUMA node][memory-issue]
+
+## Goals
+
+- Guarantee alignment of resources for isolating container's Memory and Hugepages with CPU and I/O devices.
+
+- Provide topology hint for Memory and Hugepages for Topology manager.
+
+## Non-Goals
+- Updating scheduler, pod spec is out of scope at this point.
```

Comment:
Sorry,I have a question about this. I want to know, regardless of how the kube-scheduler guarantees that the pod binding numa is successful. In my real environment, we have to consider this problem at the kube-scheduler level, that is, consider the Numa topology and binding when the kube-scheduler selects node.

Reply:
If possible, I think we can consider this problem systematically or I can help improve this part of the KEP content.

Reply:
I also agree that scheduler should be enhanced to be aware NUMA.

Unfortunately, In this KEP we want to focus on node level only.
Touching scheduler will bring complicity to this works.
For example, we will have to get a review and approval from the sig-scheduler.

But we can consider enhancing scheduler after the alpha stage of this feature.
And I think this [document](https://docs.google.com/document/d/1gPknVIOiu-c_fpLm53-jUAm-AGQQ8XC0hQYE7hq0L4c) will be helpful for you.
Kubernetes Topology-Aware Scheduling by @cbf123